dtm0/recovery.c

```diff
@@ -1677,11 +1731,40 @@ static void dtm0_restore(struct m0_fom        *fom,
 
 	do {
 		M0_SET0(&record);
-		rc = recovery_machine_log_iter_next(rf->rf_m, &rf->rf_log_iter,
-						    &rf->rf_tgt_svc, NULL,
-						    &record);
-		/* Any value except zero means that we should stop recovery. */
-		F(next) = rc == 0;
+		if (F(last_dtx_met)) {
+			/*
+			 * Last iteration reached the RECOVERING mark in the
+			 * log, don't need to send REDOs past this mark.
+			 */
+			F(next) = false;
+		} else {
+			do {
+				rc = recovery_machine_log_iter_next(
+						rf->rf_m, &rf->rf_log_iter,
+						&record);
+				/*
+				 * Any value except zero means that we should
+				 * stop recovery.
+				 *
+				 * XXX: handle cases when rc not in (0,
+				 * -ENOENT).
+				 */
+				F(next) = (rc == 0);
+				if (!F(next))
+					break;
```

Comment:
If we break from here to Line 1767, then the `rc` is not zero, and the `record` is not filled with valid data.
So, the code from Line 1768 will access an invalid `record`.

Maybe I am wrong, but please check.

Reply:
At line 1733 we set entire record to all zeroes.  Ret code can only be either 0 (and record is properly populated), or -ENOENT, in which case record is left at all zeroes.  So when we get to actual usage with non-zero RC -- record is set to all zeroes, and there is no issue.

Reply:
> So when we get to actual usage with non-zero RC -- record is set to all zeroes, and there is no issue

In this case, we will have:
```
			.dtr_payload   = record.dlr_payload,
			.dtr_txr       = record.dlr_txd,
                       .dtr_flags     =  M0_BITS(M0_DMF_EOL),
```
So, when we set the EOF flag, the dtr_payload and dtr_txr will be zeroes.
As you said, there is no issue in this case, right?

Do you think we can add your comments in the code?

Reply:
Added asserts and comments to clarify your question:

```diff
diff --git a/dtm0/recovery.c b/dtm0/recovery.c
index 824471ea..020091a2 100644
--- a/dtm0/recovery.c
+++ b/dtm0/recovery.c
@@ -1749,7 +1749,9 @@ static void dtm0_restore(struct m0_fom        *fom,
                                 * XXX: handle cases when rc not in (0,
                                 * -ENOENT).
                                 */
+                               M0_ASSERT(M0_IN(rc, (0, -ENOENT)));
                                F(next) = (rc == 0);
+                               M0_ASSERT(ergo(rc == -ENOENT, M0_IS0(&record)));
                                if (!F(next))
                                        break;
                                rc = recovery_machine_dtxid_cmp(
@@ -1760,11 +1762,21 @@ static void dtm0_restore(struct m0_fom        *fom,
                                if (participated(&record, &rf->rf_tgt_svc))
                                        break;
                                m0_dtm0_log_iter_rec_fini(&record);
+                               M0_SET0(&record);
                                if (F(last_dtx_met))
                                        F(next) = false;
                        } while (!F(last_dtx_met));
                }

+               /*
+                * At this point, a record is either fully initialized and must
+                * be sent, or is all-zero and F(next) is then set to false, so
+                * we will send EOL flag.  With current implementation, EOL
+                * flags should be sent on a separate message, which must not
+                * have any payload.
+                */
+               M0_ASSERT(equi(M0_IS0(&record), !F(next)));
+
                redo = (struct dtm0_req_fop) {
                        .dtr_msg       = DTM_REDO,
                        .dtr_initiator = F(initiator),
```